File: tests/components/subsurface/dummyfortran/dummy.f90
subroutine initialise(y, x, state_a_m1, state_b_m1)
    implicit none

    ! spaceshape
    integer, intent(in) :: y, x
    ! component states
    real(kind=8), intent(inout), dimension(y, x) :: state_a_m1, state_b_m1

    state_a_m1 = 0
    state_b_m1 = 0

end subroutine initialise

subroutine run(y, x, &
               transfer_i, transfer_n, &
               driving_a, &
               parameter_a, &
               state_a_m1, state_a_0, state_b_m1, state_b_0, &
               transfer_k, transfer_m, &
               output_x)

    implicit none

    ! spaceshape
    integer, intent(in) :: y, x
    ! from exchanger
    real(kind=8), intent(in), dimension(y, x) :: transfer_i, transfer_n
    ! component driving data
    real(kind=8), intent(in), dimension(y, x) :: driving_a
    ! component parameters
    real(kind=8), intent(in), dimension(y, x) :: parameter_a
    ! component states
    real(kind=8), intent(in), dimension(y, x) :: state_a_m1, state_b_m1
    real(kind=8), intent(inout), dimension(y, x) :: state_a_0, state_b_0
    ! to exchanger
    real(kind=8), intent(out), dimension(y, x) :: transfer_k, transfer_m
    ! component outputs
    real(kind=8), intent(out), dimension(y, x) :: output_x

    state_a_0 = state_a_m1 + 1
    state_b_0 = state_b_m1 + 2

    transfer_k = (driving_a * parameter_a) + transfer_n + state_a_0
    transfer_m = (driving_a * parameter_a) + transfer_i + state_b_0

    output_x = (driving_a * parameter_a) + transfer_n - state_a_0

end subroutine run

subroutine finalise()
    implicit none

end subroutine finalise

File: tests/tests/components/subsurface/dummyc/dummy.c
void initialise_(int ny, int nx,
                 // component states
                 double *state_a_m1, double *state_b_m1)
{
  int j, k;
  int jk;

  for (j=0; j < ny; j++)
    for (k=0; k < nx; k++)
    {
      // vectorisation of 3d-array
      jk = k + nx * j;
      // initialise states
      state_a_m1[jk] = 0.0;
      state_b_m1[jk] = 0.0;
    }
}

void run_(int ny, int nx,
          // from exchanger
          double *transfer_i, double *transfer_n,
          // component driving data
          double *driving_a,
          // component parameters
          double *parameter_a,
          // component states
          double *state_a_m1, double *state_a_0,
          double *state_b_m1, double *state_b_0,
          // to exchanger
          double *transfer_k, double *transfer_m,
          // component outputs
          double *output_x)
{
  int j, k;
  int jk;

  for (j=0; j < ny; j++)
    for (k=0; k < nx; k++)
    {
      // vectorisation of 3d-array
      jk = k + nx * j;
      // update states
      state_a_0[jk] = state_a_m1[jk] + 1;
      state_b_0[jk] = state_b_m1[jk] + 2;
      // compute transfers to exchanger
      transfer_k[jk] = (driving_a[jk] * parameter_a[jk]) + transfer_n[jk]
        + state_a_0[jk];
      transfer_m[jk] = (driving_a[jk] * parameter_a[jk]) + transfer_i[jk]
        + state_b_0[jk];
      // compute outputs
      output_x[jk] = (driving_a[jk] * parameter_a[jk]) + transfer_n[jk]
        - state_a_0[jk];
    }
}

void finalise_(void)
{}
